### PR TITLE
[GEOS-8667] WMTS Multidimensional GetHistogram is memory bound and can cause OOM over large data sets

### PR DESCRIPTION
**Pull request by aaime**
Requires the changes in https://github.com/geotools/geotools/pull/1843 to be build and get into the M2 repos for tests to pass

### CODE REVIEW

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java [hunk 1]
```diff
@@ -81,27 +94,164 @@ protected DomainSummary getDomainSummary(FeatureCollection features, String attr
...
+     * Small easter egg to allow testing, we might want to extend it to a per layer configuration
+     * in case there are large shapefile layers involved in this (computing min/max/groupby is
+     * three full data scans in that case). Or have a way to figure out if a collection can optimize
+     * out a visit (completely missing right now, that would be a significant API change).
+     * @return
+     */
+    private boolean loadDataInMemory() {
+        String value = Optional.ofNullable(GeoServerExtensions.getProperty("WMTS_HISTOGRAM_IN_MEMORY"))
+                .orElse("false");
+        return Boolean.getBoolean(value);
```
Comment (nmco) [inline src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java]: If ``value`` is ``NULL``, ``Boolean.getBoolean(value)`` will return ``FALSE`` anyway, but I understand that this is more explicit.

**Comment by aaime** [comment]
In memory yes, but streaming, instead of making GeoServer go OOM

File: src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java (hunk 1 above)
Comment (aaime) [reply src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/Dimension.java]: Good point, I can remove that optional

**Comment by aaime** [comment]
Ah, about the test, it can be done using H2, no need to make it an online one, but it would not add much, unless the code is modified in a way that allows to check if the visitor has been optimized out or not. As long as the memory visit works, it's up to GT2 to guarantee a in memory visit and a optimized sql one provide the same result.

**Comment by nmco** [comment]
Ok that makes sense, thanks for the feedback.

**Pull request merged**